google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/async_stream_pusher.rb

```diff
@@ -119,25 +127,34 @@ def stopped?
             synchronize { @stopped }
           end
 
+          def batch_nil?
+            synchronize { @batch.nil? }
+          end
+
           protected
 
           def run_background
-            synchronize do
-              until @stopped
-                if @batch.nil?
-                  @cond.wait
+            until stopped?
+              if batch_nil?
+                synchronize do
+                  @cond.wait # wait until broadcast
                   next
                 end
+              end
 
-                time_since_batch_creation = Time.now - @batch_created_at
-                if time_since_batch_creation > @interval
-                  # interval met, publish the batch...
-                  push_batch_request!
-                  @cond.wait
-                else
-                  # still waiting for the interval to publish the batch...
-                  @cond.wait(@interval - time_since_batch_creation)
-                end
+              time_until_next_push = synchronize do
+                @interval - (Time.now - @batch_created_at)
```

Comment:
This is probably a bug. AFAICT `!@batch_created_at.nil?` is not an invariant (because other methods call `push_batch_request!` before signaling), so this may throw an exception.

Reply:
What is the fix?

Reply:
You can test for nil, and assume that if it's nil there is no batch (so maybe just `continue` to go back to the first wait)

Reply:
sorry, `next`. I'm forgetting my Ruby.

Reply:
Even with a `synchronize` wrapping the loop?

Reply:
I moved this ivar to be an attribute on `@batch`, so it should never be out of sync.